### ISSUE TITLE
No HDMI signal from the Jetson Nano

### ISSUE THREAD

**Issue by codethief** [balena-io-experimental/jetson-sample-new#3]
I followed the instructions in the [blog post](https://www.balena.io/blog/how-to-use-nvidia-jetson-devices-on-balenaos/) and, [having fixed the Dockerfile](https://github.com/balena-io-playground/jetson-nano-sample-new/pull/2), I am now able to ssh into the cuda and opencv containers and fire up the X server and the test scripts. However, when I connect my monitor via HDMI, the screen complains that there is no incoming HDMI signal. Meanwhile, the X server seems to be running fine and no errors are shown in the logs. This happens for *both* the cuda and the opencv container. I should mention, I'm new to BalenaOS and this is the very first time I'm trying to run an X server on the platform, so while I can attest to the fact that my Nano's HDMI worked fine when I was using Nvidia's Ubuntu image, I haven't run X successfully on an BalenaOS-equipped Jetson Nano yet. (Meaning that from my POV the issue could either be due to the involvement of L4T in the containers here or it could be related to the X server itself.)

Anyway, I'm really at a loss here and would be grateful for any pointers.

FWIW, the output of xrandr while my screen is connected looks ok:

```
root@8546bd40d32f:/usr/src/app# xrandr -d :0
Screen 0: minimum 8 x 8, current 640 x 480, maximum 16384 x 16384
HDMI-0 connected primary (normal left inverted right x axis y axis)
   2560x1440     59.96 +
   2048x1080     60.00    24.00  
   1920x1080     60.00    59.95    50.00  
   1600x1200     60.01  
   1280x1024     75.03    60.00  
   1280x720      60.00    59.94    50.00  
   1152x864      75.00  
   1024x768      75.03    60.01  
   800x600       75.00    60.32  
   720x576       50.00  
   720x480       59.94  
   720x400       70.04  
   640x480       75.00    59.94    59.94  
DP-0 disconnected (normal left inverted right x axis y axis)
```

**Comment by codethief** [comment]
Turns out the monitor must already be connected when starting the X server. Only then will it yield an HDMI signal. (I didn't notice this initially because I only have one monitor connected to my laptop, so I was switching back and forth.)

I am leaving this ticket open, though, because IMO the screen should not have to be connected when firing up the X server. In particular, the X server on Nvidia's Ubuntu image does not require this. Whether or not this is fixable I don't know but it should at least be noted in the blog post.

FWIW the same issue seems to happen with https://github.com/balenablocks/browser (after adapting it slightly to run on the Nano). So this seems to be a more general issue and maybe other people struggling like me will find this ticket here useful.

**Comment by codethief** [comment]
I have analyzed the issue a bit further now and noticed that 

1. *Sometimes* connecting the monitor while the X server is already running works without issues, but most of the time not. `dmesg` reflects this as follows:

```
# Output of dmesg when connecting the monitor via HDMI. This time the monitor ends up receiving a signal.

[28945.629017] tegradc tegradc.0: blank - powerdown
[28945.629107] tegradc tegradc.0: unblank
[28945.676939] tegradc tegradc.0: nominal-pclk:241500000 parent:241500000 div:1.0 pclk:241500000 239085000~263235000
[28945.677521] tegradc tegradc.0: hdmi: tmds rate:241500K prod-setting:prod_c_hdmi_150m_300m
[28945.679726] tegradc tegradc.0: hdmi: get RGB quant from EDID.
[28945.679763] tegradc tegradc.0: hdmi: get YCC quant from EDID.
[28945.717202] extcon-disp-state extcon:disp-state: cable 47 state 1
[28945.717215] Extcon AUX1(HDMI) enable
[28945.773928] extcon-disp-state extcon:disp-state: cable 51 state 1
[28945.787274] Extcon HDMI: HPD enabled
[28945.787766] tegradc tegradc.0: hdmi: plugged
[28945.818997] tegradc tegradc.0: blank - powerdown
[28945.880569] extcon-disp-state extcon:disp-state: cable 47 state 0
[28945.880572] Extcon AUX1(HDMI) disable
[28945.915659] tegradc tegradc.0: unblank
[28945.929460] tegradc tegradc.0: nominal-pclk:241500000 parent:241500000 div:1.0 pclk:241500000 239085000~263235000
[28945.929581] tegradc tegradc.0: hdmi: tmds rate:241500K prod-setting:prod_c_hdmi_150m_300m
[28945.930776] tegradc tegradc.0: hdmi: get RGB quant from EDID.
[28945.930786] tegradc tegradc.0: hdmi: get YCC quant from EDID.
[28945.965801] extcon-disp-state extcon:disp-state: cable 47 state 1
[28945.965804] Extcon AUX1(HDMI) enable
[28946.009434] tegradc tegradc.0: unblank
[28946.013219] tegradc tegradc.1: blank - powerdown
```


```
# Output of dmesg when connecting the monitor via HDMI. This time the monitor ends up NOT receiving a signal.

[29096.295198] tegradc tegradc.0: blank - powerdown
[29096.295285] tegradc tegradc.0: unblank
[29096.377404] tegradc tegradc.0: nominal-pclk:241500000 parent:241500000 div:1.0 pclk:241500000 239085000~263235000
[29096.377720] tegradc tegradc.0: hdmi: tmds rate:241500K prod-setting:prod_c_hdmi_150m_300m
[29096.379147] tegradc tegradc.0: hdmi: get RGB quant from EDID.
[29096.379163] tegradc tegradc.0: hdmi: get YCC quant from EDID.
[29096.414742] extcon-disp-state extcon:disp-state: cable 47 state 1
[29096.414748] Extcon AUX1(HDMI) enable
[29096.429789] tegradc tegradc.0: sync windows ret = 246
[29096.483765] extcon-disp-state extcon:disp-state: cable 51 state 1
[29096.489951] Extcon HDMI: HPD enabled
[29096.490114] tegradc tegradc.0: hdmi: plugged
[29096.514407] tegradc tegradc.0: blank - powerdown
[29096.580029] extcon-disp-state extcon:disp-state: cable 47 state 0
[29096.580034] Extcon AUX1(HDMI) disable
[29096.615788] tegradc tegradc.1: blank - powerdown
```


2. Another realization: I don't necessarily *have* to have the monitor connected via HDMI when starting the X server but, as a workaround, I can also "activate" the monitor later on using `xrandr`:

```
# Situation: Monitor not connected, X server is running

root@9d9673522c11:/usr/src/app# xrandr
Screen 0: minimum 8 x 8, current 640 x 480, maximum 16384 x 16384
HDMI-0 disconnected primary (normal left inverted right x axis y axis)
DP-0 disconnected (normal left inverted right x axis y axis)


# Now I connect the monitor -> "No Signal"

root@9d9673522c11:/usr/src/app# xrandr
Screen 0: minimum 8 x 8, current 640 x 480, maximum 16384 x 16384
HDMI-0 connected primary (normal left inverted right x axis y axis)
   2560x1440     59.96 +
   2048x1080     60.00    24.00  
   1920x1080     60.00    59.95    50.00  
   1600x1200     60.01  
   1280x1024     75.03    60.00  
   1280x720      60.00    59.94    50.00  
   1152x864      75.00  
   1024x768      75.03    60.01  
   800x600       75.00    60.32  
   720x576       50.00  
   720x480       59.94  
   720x400       70.04  
   640x480       75.00    59.94    59.94  
DP-0 disconnected (normal left inverted right x axis y axis)


# Now I enable the monitor:

root@9d9673522c11:/usr/src/app# xrandr --output HDMI-0 --auto  # -> Monitor is now working!


# …and turn it off again:

root@9d9673522c11:/usr/src/app# xrandr --output HDMI-0 --off  # -> turns off

root@9d9673522c11:/usr/src/app# xrandr
Screen 0: minimum 8 x 8, current 8 x 8, maximum 16384 x 16384
HDMI-0 connected primary (normal left inverted right x axis y axis)
   2560x1440     59.96 +
   2048x1080     60.00    24.00  
   1920x1080     60.00    59.95    50.00  
   1600x1200     60.01  
   1280x1024     75.03    60.00  
   1280x720      60.00    59.94    50.00  
   1152x864      75.00  
   1024x768      75.03    60.01  
   800x600       75.00    60.32  
   720x576       50.00  
   720x480       59.94  
   720x400       70.04  
   640x480       75.00    59.94    59.94  
DP-0 disconnected (normal left inverted right x axis y axis)


# …and on again:

root@9d9673522c11:/usr/src/app# xrandr --output HDMI-0 --auto  # -> turns on

root@9d9673522c11:/usr/src/app# xrandr
Screen 0: minimum 8 x 8, current 2560 x 1440, maximum 16384 x 16384
HDMI-0 connected primary 2560x1440+0+0 (normal left inverted right x axis y axis) 600mm x 340mm
   2560x1440     59.96*+
   2048x1080     60.00    24.00  
   1920x1080     60.00    59.95    50.00  
   1600x1200     60.01  
   1280x1024     75.03    60.00  
   1280x720      60.00    59.94    50.00  
   1152x864      75.00  
   1024x768      75.03    60.01  
   800x600       75.00    60.32  
   720x576       50.00  
   720x480       59.94  
   720x400       70.04  
   640x480       75.00    59.94    59.94  
DP-0 disconnected (normal left inverted right x axis y axis)
```

FYI Activating the monitor using `xrandr --output HDMI-0 --auto` causes `dmesg` to output the following:

```
[29192.631596] tegradc tegradc.0: blank - powerdown
[29192.636504] tegradc tegradc.0: unblank
[29192.662443] tegradc tegradc.0: nominal-pclk:241500000 parent:241500000 div:1.0 pclk:241500000 239085000~263235000
[29192.662694] tegradc tegradc.0: hdmi: tmds rate:241500K prod-setting:prod_c_hdmi_150m_300m
[29192.665009] tegradc tegradc.0: hdmi: get RGB quant from EDID.
[29192.665027] tegradc tegradc.0: hdmi: get YCC quant from EDID.
[29192.700690] extcon-disp-state extcon:disp-state: cable 47 state 1
[29192.700696] Extcon AUX1(HDMI) enable
[29192.760586] tegradc tegradc.0: unblank
[29192.765346] tegradc tegradc.1: blank - powerdown
```

**Comment by codethief** [comment]
After further investigation I think I might have found the reason: I noticed that `udevadm monitor --udev` does not output any udev events when hot-plugging my monitor, meaning that udev is probably not notified by the Nvidia driver. (I heard that one could maybe set `options nvidia-drm.modeset=1` in the modprobe config but I suppose this requires a change of the BalenaOS base image, so I can't do this here.)

I did come up with a workaround, though: Turns out `udevadm monitor --kernel` does output some HDMI-related events. So I added the following background task to my Docker container's `start.sh` script:

```bash
udevadm monitor --kernel | grep --line-buffered "/devices/virtual/switch/hdmi " | while read -r line; do
    echo "HDMI monitor hot-plugging detected, executing xrandr --auto..."
    xrandr --auto
done &
```

and thinks now work perfectly!
